File: Data Structures & Algorithms/eating-bananas/submission-1.cpp
class Solution {
private:
    bool canEatPiles(const vector<int>& piles, int rate, int h) {
        int count = 0;
        
        for (const auto& pile: piles) {
            count += (pile + rate - 1) / rate;
        }

        return count <= h;
    }

public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int upper = *std::max_element(piles.begin(), piles.end());
        int lower = 1;
        int result = upper;
        
        while (lower <= upper) {
            int mid = (lower + upper) / 2;
            if (canEatPiles(piles, mid, h)) {
                upper = mid - 1;
                result = mid;
            } else {
                lower = mid + 1;
            }
        }

        return result;
    }
};

File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        i, j = 0, len(s) - 1
        s_copy = ""

        for ch in s:
            if ch.isalnum(): s_copy += ch.lower()
        
        i, j = 0, len(s_copy) - 1

        while i < j:
            if s_copy[i] != s_copy[j]: return False
            i += 1
            j -= 1
        
        return True 